### PR TITLE
HHH-14828 Bytecode enhancement generates invalid bytecode for final fields

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net/browse/HHH-14828

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/PersistentAttributeTransformer.java [hunk 1]
```diff
@@ -187,15 +192,19 @@ private boolean isEnhanced(String owner, String name, String desc) {
...
-							Visibility.PUBLIC
-					)
-					.withParameters( enhancedField.getType().asErasure() )
-					.intercept( fieldWriter( enhancedField ) );
+					.intercept( fieldReader( enhancedField ) );
+			// Final fields will only be written to from the constructor,
+			// so there's no point trying to replace final field writes with a method call.
+			if ( !enhancedField.getFieldDescription().isFinal() ) {
+				builder = builder
+						.defineMethod(
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/PersistentAttributeTransformer.java:229]: Do you rememer if Panache is taking the existance of such methods for granted?

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/PersistentAttributeTransformer.java [hunk 2]
```diff
@@ -42,12 +43,29 @@
...
 import net.bytebuddy.matcher.ElementMatcher.Junction;
 import net.bytebuddy.pool.TypePool;
+import net.bytebuddy.utility.OpenedClassReader;
 
 final class PersistentAttributeTransformer implements AsmVisitorWrapper.ForDeclaredMethods.MethodVisitorWrapper {
 
 	private static final CoreMessageLogger log = CoreLogging.messageLogger( PersistentAttributeTransformer.class );
 
 	private static final Junction<MethodDescription> NOT_HIBERNATE_GENERATED = not( nameStartsWith( "$$_hibernate_" ) );
+	private static final ModifierContributor.ForField REMOVE_FINAL_MODIFIER = new ModifierContributor.ForField() {
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/PersistentAttributeTransformer.java:53]: Interesting, I'm learning a lot here :)

**Review by Sanne — APPROVED** [review]
Excellent, thanks!

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/PersistentAttributeTransformer.java (hunk 1 above)
Comment (yrodiere) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/PersistentAttributeTransformer.java:229]: It is. However, I think Panache only performs extended enhancement, and I think this enhancement supposedly avoids transforming field access within the class declaring the field?

Not sure if it's trying or succeeding, though; ORM is trying but failing, as far as I can see, but that's another problem.

In any case, we'll have to add tests in Quarkus. I'll submit a PR.

**Pull request merged**